lib/ansible/module_utils/vmware.py

```diff
@@ -0,0 +1,181 @@
+    try:
+        host_folder = datacenter.hostFolder
+        for folder in host_folder.childEntity:
+            if folder.name == cluster_name:
+                return folder
+        return None
+    # This exception should be handled in the module that calls this method
+    # and fail with an appropriate message to module.fail_json()
+    except vmodl.MethodFault:
+        raise
+
+
+def find_datacenter_by_name(content, datacenter_name, throw=True):
+    try:
+        datacenters = get_all_objs(content, [vim.Datacenter])
+        for dc in datacenters:
+            if dc.name == datacenter_name:
+                return dc
+
+        return None
+    # This exception should be handled in the module that calls this method
+    # and fail with an appropriate message to module.fail_json()
+    except vmodl.MethodFault:
+        raise
+
+
+def find_dvs_by_name(content, switch_name):
+    try:
+        vmware_distributed_switches = get_all_objs(content, [vim.dvs.VmwareDistributedVirtualSwitch])
+        for dvs in vmware_distributed_switches:
+            if dvs.name == switch_name:
+                return dvs
+        return None
+    # This exception should be handled in the module that calls this method
+    # and fail with an appropriate message to module.fail_json()
+    except vmodl.MethodFault:
+        raise
+
+
+def find_hostsystem_by_name(content, hostname):
+    try:
+        host_system = get_all_objs(content, [vim.HostSystem])
+        for host in host_system:
+            if host.name == hostname:
+                return host
+        return None
+    # This exception should be handled in the module that calls this method
+    # and fail with an appropriate message to module.fail_json()
+    except vmodl.MethodFault:
+        raise
+
+
+def vmware_argument_spec():
+    return dict(
+        hostname=dict(type='str', required=True),
+        username=dict(type='str', aliases=['user', 'admin'], required=True),
+        password=dict(type='str', aliases=['pass', 'pwd'], required=True, no_log=True),
+    )
+
+
+def connect_to_api(module, disconnect_atexit=True):
+    hostname = module.params['hostname']
+    username = module.params['username']
+    password = module.params['password']
+    try:
+        service_instance = connect.SmartConnect(host=hostname, user=username, pwd=password)
+
+        # Disabling atexit should be used in special cases only.
+        # Such as IP change of the ESXi host which removes the connection anyway.
+        # Also removal significantly speeds up the return of the module
+
+        if disconnect_atexit:
+            atexit.register(connect.Disconnect, service_instance)
+        return service_instance.RetrieveContent()
+    except vim.fault.InvalidLogin as invalid_login:
+        module.fail_json(msg=invalid_login.msg)
+    except requests.ConnectionError:
+        module.fail_json(msg="Unable to connect to vCenter or ESXi API on TCP/443.")
```

Comment:
it is useful to also add the api exception info:
module.fail_json(msg="Unable to connect to vCenter or ESXi API on TCP/443.", apierror=str(e))